### PR TITLE
Seth cleanup

### PR DESCRIPTION
**Pull request by Sezzary**
Needs testing. In particular, check whether effects and the kill animation work properly.

### CODE REVIEW

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp [hunk 1]
```diff
@@ -191,89 +160,87 @@ namespace TEN::Entities::TR4
...
-							item->Pose.Position.y += SECTOR(1.5f);
-							item->Animation.TargetState = SETH_STATE_JUMP_AIR_SHOOT;
+							item->Pose.Position.y += BLOCK(3 / 2.0f);
+							item->Animation.TargetState = SETH_STATE_JUMP_PROJECTILE_ATTACK;
 							item->ItemFlags[0] = 0;
 						}
 						else
 						{
-							item->Pose.Position.y += SECTOR(1.5f);
+							item->Pose.Position.y += BLOCK(3 / 2.0f);
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Entity/tr4_setha.cpp:220]: As above.

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp [hunk 2]
```diff
@@ -494,182 +467,169 @@ namespace TEN::Entities::TR4
...
 
 		switch (item->Animation.ActiveState)
 		{
-		case SETH_STATE_SHOOT_RIGHT_AND_LEFT:
-		case SETH_STATE_FLY_AIR_SHOOT:
-			if (item->ItemFlags[0] < 78 && (GetRandomControl() & 0x1F) < item->ItemFlags[0])
+		case SETH_STATE_DUAL_PROJECTILE_ATTACK:
+		case SETH_STATE_HOVER_PROJECTILE_ATTACK:
+			if (item->ItemFlags[0] < 78 &&
+				(GetRandomControl() & 0x1F) < item->ItemFlags[0])
```
Comment (Lwmte) [inline TombEngine/Objects/TR4/Entity/tr4_setha.cpp:518]: Why this GetRandomControl was not replaced?

File: TombEngine/Game/effects/item_fx.cpp [hunk 3]
```diff
@@ -21,10 +21,11 @@ namespace TEN::Effects::Items
 		item->Effect.Type = EffectType::Fire;
 		item->Effect.Count = timeout;
 		item->Effect.LightColor = Vector3(0.8f, 0.5f, 0.0f);
-		item->Effect.PrimaryEffectColor = item->Effect.SecondaryEffectColor = Vector3::Zero;
+		item->Effect.PrimaryEffectColor = Vector3::Zero;
```
[the remark this answers is not in the document]
Comment (Sezzary) [reply TombEngine/Game/effects/item_fx.cpp:24]: I find myself tripping on these when reading. They could be done like this at least:
```
item->Effect.PrimaryEffectColor =
item->Effect.SecondaryEffectColor = Vector3::Zero;
```

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp [hunk 4]
```diff
@@ -177,7 +146,7 @@ namespace TEN::Entities::TR4
 		int height3 = GetCollision(x, y, z, item->RoomNumber).Position.Floor;
 
 		bool canJump = false;
-		if ((y < (height1 - CLICK(1.5f)) || y < (height2 - CLICK(1.5f))) &&
+		if ((y < (height1 - CLICK(3 / 2.0f)) || y < (height2 - CLICK(3 / 2.0f))) &&
```
[the remark this answers is not in the document]
Comment (Sezzary) [reply TombEngine/Objects/TR4/Entity/tr4_setha.cpp:149]: I've been doing it like this for a while now. :P

File: TombEngine/Game/effects/effects.cpp [hunk 5]
```diff
@@ -1718,29 +1718,29 @@ void TriggerAttackFlame(const Vector3i& pos, const Vector3& color, int size)
...
-	spark.xVel = (GetRandomControl() & 0xFF) - 128;
+	spark.z = pos.z + Random::GenerateInt(-8, 8);
+	spark.xVel = Random::GenerateInt(-128, 128);
 	spark.yVel = 0;
-	spark.zVel = (GetRandomControl() & 0xFF) - 128;
+	spark.zVel = Random::GenerateInt(-128, 128);
 	spark.friction = 5;
 	spark.flags = SP_EXPDEF | SP_DEF | SP_SCALE;
-	spark.rotAng = GetRandomControl() & 0xFFF;
+	spark.rotAng = Random::GenerateAngle(0, ANGLE(22.5f));
```
[the remark this answers is not in the document]
Comment (Sezzary) [reply TombEngine/Game/effects/effects.cpp:1733]: Oh yeah, I forgot about this. Will change it back for now.

File: TombEngine/Game/effects/effects.cpp [hunk 6]
```diff
@@ -1718,29 +1718,29 @@ void TriggerAttackFlame(const Vector3i& pos, const Vector3& color, int size)
 	spark.dG = color.y;
 	spark.dB = color.z;
 	spark.fadeToBlack = 8;
-	spark.colFadeSpeed = (GetRandomControl() & 3) + 4;
+	spark.colFadeSpeed = Random::GenerateInt(4, 4);
```
[the remark this answers is not in the document]
Comment (Sezzary) [reply TombEngine/Game/effects/effects.cpp:1721]: Will change to [4, 8] range.

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp [hunk 7]
```diff
@@ -678,4 +638,58 @@ namespace TEN::Entities::TR4
...
+
+		fx.pos.Position.x = pose.Position.x;
+		fx.pos.Position.y = pose.Position.y - Random::GenerateInt(-32, 32);
+		fx.pos.Position.z = pose.Position.z;
+
+		fx.pos.Orientation.x = pose.Orientation.x;
+		fx.pos.Orientation.y = pose.Orientation.y;
+		fx.pos.Orientation.z = 0;
+		fx.roomNumber = roomNumber;
+		fx.counter = (GetRandomControl() * 2) - ANGLE(180.0f); // TODO: This isn't an angle.
```
[the remark this answers is not in the document]
Comment (Sezzary) [reply TombEngine/Objects/TR4/Entity/tr4_setha.cpp:658]: I don't get what's happening here. The range appears to be [-32768, 32767]?

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp (hunk 2 above)
Comment (Sezzary) [reply TombEngine/Objects/TR4/Entity/tr4_setha.cpp:518]: I suppose it generates an int in the range [0, 32], but I don't know what `ItemFlags[0]` contains exactly.

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (Lwmte) [reply TombEngine/Objects/TR4/Entity/tr4_setha.cpp:658]: Yes, seems so.

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp (hunk 2 above)
Comment (Lwmte) [reply TombEngine/Objects/TR4/Entity/tr4_setha.cpp:518]: From the looks of it, it seems like some kind of AI modifier, kind of "frustration" with the attack. When `ItemFlags[0]` reaches certain value, a decision is made on certain kinds of attacks, after which his "frustration" is "resolved" (value is reset to 0).

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Lwmte) [reply TombEngine/Objects/TR4/Entity/tr4_setha.cpp:149]: I think we already discussed this on Discord and decided on that fractions should be only used on probabilities. `CLICK(3 / 2.0f)` is harder to read than `CLICK(1.5f)`. So I think you should not continue with converting height values to this format.

File: TombEngine/Game/effects/item_fx.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Lwmte) [reply TombEngine/Game/effects/item_fx.cpp:24]: You are right, in fact mutiple one-line assignment is considered bad practice by a lot of people. So let's keep it as is.

File: TombEngine/Game/effects/effects.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Lwmte) [reply TombEngine/Game/effects/effects.cpp:1733]: Thanks :)

File: TombEngine/Objects/TR4/Entity/tr4_setha.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sezzary) [reply TombEngine/Objects/TR4/Entity/tr4_setha.cpp:149]: I love fractions and I am sad, but okay. :')

**Comment by Nemoel-Tomo** [comment]
I have tested it today with seth and harphy and I have found no bugs or errors so far.

**Pull request merged**